### DISCUSSION TITLE
Podman pull does not prefer zstd:chunked images

### DISCUSSION THREAD

**Discussion by p5** [containers/podman#23140] [General]
According to [RHDan's forum post](https://discussion.fedoraproject.org/t/switch-fedora-container-images-to-support-zstd-chunked-format-by-default/123712/8), the current container tools are smart enough to favour pulling zstd:chunked images from registries when in a manifest alongside gzipped images.  Even if the gzipped image appears first.

Over at Universal Blue, we have started looking to push gzip and zstd:chunked alongside eachother, however are facing some strange behaviour when pulling or inspecting the images.

Our images:
- [Manifest: `ghcr.io/ublue-os/bluefin:40-testing`](https://github.com/ublue-os/bluefin/pkgs/container/bluefin/236972929?tag=40-testing)
- Consists of:
  - [Gzip: `ghcr.io/ublue-os/bluefin@sha256:a78fe8009beeadb657bad6e76aa47e99b5b5945e618c3192b0ab4f77a06345ef`](https://github.com/ublue-os/bluefin/pkgs/container/bluefin/236971677?tag=testing-gzip)
  - [Zstd:chunked: `ghcr.io/ublue-os/bluefin@sha256:86d32d18a0eb7329293142bb0948b03e9a3896cde004f6faef0c60c3baac9375`](https://github.com/ublue-os/bluefin/pkgs/container/bluefin/236972924?tag=testing-zstd-chunked)

When we target the manifest **tag** for during a `podman pull`, we receive the gzipped image where we would have expected the zstd image.
When we target the manifest **digest** during a `skopeo inspect`, we receive information about the zstd:chunked image, which is expected.

How can we ensure that users pull the chunked image when running with podman, while keeping compatibility with older tooling which does not support zstd:chunked?

We're not quite sure what we're missing, so any assistance would be greatly appreciated.

Thank you
Rob

**Comment by vrothberg** [comment]
I think we missed mentioning that the zstd-compressed image needs to be marked with annotation in the OCI index/manifest list:
```json
{
    "mediaType": "application/vnd.oci.image.manifest.v1+json",
    "digest": "sha256:bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb",
    "size": 772,
    "annotations": {
        "io.github.containers.compression.zstd": "true"
    },
    "platform": {
        "architecture": "amd64",
        "os": "linux"
    }
},
```

**Reply by p5** [reply]
Thanks!  That makes a lot of sense.

Noticed something strange though.  After successfully adding that annotation to the manifest list locally (see below), I push the manifest to GHCR and run a `podman manifest inspect` but these annotations are lost.

Do you happen to know if there's some incompatibility with GHCR?

Manifest (local):

<details>

```json
{
    "schemaVersion": 2,
    "mediaType": "application/vnd.docker.distribution.manifest.list.v2+json",
    "manifests": [
        {
            "mediaType": "application/vnd.oci.image.manifest.v1+json",
            "size": 10730,
            "digest": "sha256:852c0aead11e5f2de95b362c7ca4d7a2aaf9457e7d4d7aaf408a19388df9efa6",
            "platform": {
                "architecture": "amd64",
                "os": "linux"
            }
        },
        {
            "mediaType": "application/vnd.oci.image.manifest.v1+json",
            "size": 39250,
            "digest": "sha256:28c0c0dd0b4a25d615617e15f3a4b6c5372104c6562a9e2f15138ff2114acb2e",
            "platform": {
                "architecture": "amd64",
                "os": "linux"
            },
            "annotations": {
                "io.github.containers.compression.zstd": "true"
            }
        }
    ]
}
```

</details>

Manifest (remote):

<details>

```json
{
    "schemaVersion": 2,
    "mediaType": "application/vnd.docker.distribution.manifest.list.v2+json",
    "manifests": [
        {
            "mediaType": "application/vnd.oci.image.manifest.v1+json",
            "size": 10730,
            "digest": "sha256:852c0aead11e5f2de95b362c7ca4d7a2aaf9457e7d4d7aaf408a19388df9efa6",
            "platform": {
                "architecture": "amd64",
                "os": "linux"
            }
        },
        {
            "mediaType": "application/vnd.oci.image.manifest.v1+json",
            "size": 39250,
            "digest": "sha256:28c0c0dd0b4a25d615617e15f3a4b6c5372104c6562a9e2f15138ff2114acb2e",
            "platform": {
                "architecture": "amd64",
                "os": "linux"
            }
        }
    ]
}
```

</details>

I completely understand knowing compatibility of different services isn't really your area, but just wondering if you happen to have heard of anything.

**Reply by p5** [reply]
Think this is an unrelated issue.  
`podman manifest inspect` refuses to show annotations for remote manifests.  Will make an issue as it stumped me for hours!

**Reply by queeup** [reply]
Can it be related to [annotation level?](https://docs.docker.com/build/ci/github-actions/annotations/#configure-annotation-level)

**Reply by p5** [reply]
Not sure.  
I'm just running `podman manifest add --annotation=xyz=abc <manifest> <image>`, rather than using the metadata action.
The annotations are available when inspecting the local manifest, but not when inspecting the same manifest from a remote registry.

**Reply by vrothberg** [reply]
@nalind @giuseppe can you take a look?

**Reply by queeup** [reply]
podman manifest push?

**Reply by p5** [reply]
> podman manifest push?

Yes.  I am running `podman manifest create`, `podman manifest add --annotation ...` and `podman manifest push`.
The annotations are not visible on the pushed manifest when running `podman manifest inspect`, but they are visible on the local manifest.

Version: 5.1.1

**Reply by giuseppe** [reply]
it works fine for me.  I've created the manifest with the following commands:

```
podman manifest create quay.io/giuseppe/zstd-chunked:fedora-manifest
podman manifest add quay.io/giuseppe/zstd-chunked:fedora-manifest quay.io/giuseppe/zstd-chunked:fedora-gzip
podman manifest add --annotation io.github.containers.compression.zstd=true quay.io/giuseppe/zstd-chunked:fedora-manifest quay.io/giuseppe/zstd-chunked:fedora
```

where `quay.io/giuseppe/zstd-chunked:fedora-gzip` is a gzip compressed image, and `quay.io/giuseppe/zstd-chunked:fedora` is using zstd:chunked.

and then I've:

```
$ skopeo inspect --raw docker://quay.io/giuseppe/zstd-chunked:fedora-manifest | jq .
{
  "schemaVersion": 2,
  "mediaType": "application/vnd.oci.image.index.v1+json",
  "manifests": [
    {
      "mediaType": "application/vnd.oci.image.manifest.v1+json",
      "digest": "sha256:8754d19a0335a0279efeac7621f9fc52ce094763fa7ed620ef98a48be3441a2c",
      "size": 615,
      "platform": {
        "architecture": "amd64",
        "os": "linux"
      }
    },
    {
      "mediaType": "application/vnd.oci.image.manifest.v1+json",
      "digest": "sha256:9383f2cde72625a421683a1d75fc5b1067473479c119f36791f0b203eebaa2f3",
      "size": 922,
      "annotations": {
        "io.github.containers.compression.zstd": "true"
      },
      "platform": {
        "architecture": "amd64",
        "os": "linux"
      }
    }
  ]
}

```

**Reply by p5** [reply]
The possible issue in question was with `podman manifest inspect`, which I would expect to show the same information as skopeo, but doesn't.

Apologies if the thread didn't get the problem across well.  The manifests are created perfectly, but confusion happens when you inspect those manifests after pushing them.

```json
➜  ~ podman manifest inspect quay.io/giuseppe/zstd-chunked:fedora-manifest
{
    "schemaVersion": 2,
    "mediaType": "application/vnd.docker.distribution.manifest.list.v2+json",
    "manifests": [
        {
            "mediaType": "application/vnd.oci.image.manifest.v1+json",
            "size": 615,
            "digest": "sha256:8754d19a0335a0279efeac7621f9fc52ce094763fa7ed620ef98a48be3441a2c",
            "platform": {
                "architecture": "amd64",
                "os": "linux"
            }
        },
        {
            "mediaType": "application/vnd.oci.image.manifest.v1+json",
            "size": 922,
            "digest": "sha256:9383f2cde72625a421683a1d75fc5b1067473479c119f36791f0b203eebaa2f3",
            "platform": {
                "architecture": "amd64",
                "os": "linux"
            }
        }
    ]
}
```

**Reply by p5** [reply]
Have created an issue which hopefully explains this better
https://github.com/containers/podman/issues/23163

**Comment by p5** [comment]
Will close this since Podman does prefer zstd:chunked images when the required annotations are added to the manifest
